src/sync.h

```diff
@@ -398,4 +398,104 @@ class CSemaphoreGrant
     }
 };
 
+/**
+ * A smart container that protects an owned object with a mutex.
+ * This is like a mutex with an extension - owning an object of arbitrary type
+ * and providing a dereference operation which gives access to the owned object
+ * and transparently locks the mutex for the duration of the access.
+ * For example, the following is safe:
+ * @code{.cpp}
+ * Synced<std::vector<int>> v;
+ *
+ * std::thread t1{[&v]() { v->push_back(5); }};
+ * std::thread t2{[&v]() { v->push_back(6); }};
+ * @endcode
+ *
+ * It is possible so acquire the mutex for longer time than just for a single call:
+ * @code{.cpp}
+ * Synced<std::vector<int>> v{5, 6, 7, 8};
+ *
+ * // v is locked for the entire duration of the loop
+ * for (auto& i : *v) {
+ *     i *= 10;
+ * }
+ *
+ * {
+ *     LOCK(v);
+ *     size_t size = v->size();
+ *     // The container will not be changed by another thread between the two size() calls.
+ *     assert(size == v->size());
+ * }
+ * @endcode
+ */
+template <typename T>
+class Synced : public RecursiveMutex
+{
+public:
+    /**
+     * Construct the smart container and its internal object of type T,
+     * passing `args...` to T's constructor.
+     */
+    template <typename... Args>
+    Synced(Args... args) : m_obj{args...}
+    {
+    }
+
+    /**
+     * An auxiliary class that holds the parent locked for the duration of its
+     * lifetime and provides access to the internal object.
+     */
+    class Proxy : ::UniqueLock<RecursiveMutex>
+    {
+    public:
+        Proxy(Synced& parent, const char* mutex_name, const char* file_name, int line)
+            : ::UniqueLock<RecursiveMutex>{parent, mutex_name, file_name, line},
+              m_obj_ref{parent.m_obj}
+        {
+        }
+
+        T* operator->() { return &m_obj_ref; }
+
+        template <typename Key>
+        auto& operator[](Key key)
+        {
+            return m_obj_ref[key];
+        }
+
+        auto begin() { return m_obj_ref.begin(); }
+
+        auto end() { return m_obj_ref.end(); }
+
+    private:
+        T& m_obj_ref;
+    };
+
+    /**
+     * Access `m_obj`, holding `this` locked for the duration of the access.
+     * This causes a chain of `operator->()` calls:
+     * 1. `Synced::operator->()` is called, which returns `Proxy` (the
+     * constructor of `Proxy` locks `this`), then
+     * 2. `Proxy::operator->()` is called, which returns a raw pointer to
+     * `m_obj` which is dereferenced in order to call the relevant method.
+     */
+    Proxy operator->()
+    {
+        return Proxy{*this, "Synced::operator->()", __FILE__, __LINE__};
```

Comment:
Yeah, me too. Clang keeps emitting bogus warnings about loops like `for (auto& a : *obj)` for everything I tried, so I gave up. It has its limitations after all.

Reply:
Oh!! How about this:

```c++
template<typename T>
struct Guarded
{
private:
    mutable Mutex m;
    T obj GUARDED_BY(m);
public:
    class SCOPED_LOCKABLE Proxy
    {
    private:
        Guarded& g;
    public:
        Proxy(Guarded& g) EXCLUSIVE_LOCK_FUNCTION(g.m) : g{g}
        {
            g.m.lock();
        }
        ~Proxy() UNLOCK_FUNCTION() { g.m.unlock(); }

        T& operator*() { return g.obj; }
        T* operator->() { return &g.obj; }
    };
};

#define PROXY(name, guarded) decltype(guarded)::Proxy name{guarded};

Guarded<uint32_t> global_i;
Guarded<std::vector<uint32_t>> global_nums;
void foo()
{
    PROXY(p, global_i);
    *p = 5;
    *p += 6;
    // PROXY(p2, global_i); // error: acquiring mutex 'global_i.m' that is already held [-Werror,-Wthread-safety-analysis]
    // *p2 = 9;
}

int64_t lock_already_held(Guarded<std::vector<uint32_t>>::Proxy &p)
{
    int64_t j = 0;
    for (auto i : *p) { j += i; }
    return j;
}

int64_t bar()
{
    PROXY(p, global_nums);
    int64_t j = 0;
    for (auto i : *p) { j += i; }
    return j + lock_already_held(p);
}
```

That:

 * gives compile time errors if you try accessing without a lock
 * doesn't require recursive locks
 * gives clang compile time errors if you try double locking

which hits everything I wanted, and doesn't even seem too klunky. Perhaps needs some additional work to handle `const`.

Reply:
I think the macro is necessary -- otherwise `return Proxy{*this}` means you have to annotate the `Proxy(Proxy&&)` move constructor, but the only thing you can annotate that with is locking its internal reference to `g.m` but that then creates an alias of the mutex which prevents clang from detecting when double locking is happening since clang can't tell that two different invocations are locking aliases of the same lock.

Changing the name and the argument order makes lots of sense, I just posted it as soon as I found something that worked. `Guarded<T>` probably isn't a great name either. (I think it makes sense for `Proxy` to match whatever the macro is named)

Reply:
What about this:

```cpp
template <typename T>
class Synced
{
public:
    template <typename... Args>
    Synced(Args... args) : m_obj{args...}
    {
    }

    class SCOPED_LOCKABLE Proxy : ::UniqueLock<Mutex>
    {
    public:
        Proxy(Synced& parent) EXCLUSIVE_LOCK_FUNCTION(parent.m_mutex)
            : ::UniqueLock<Mutex>{parent.m_mutex, "...", "...", 123}, m_parent{parent}
        {
        }

        ~Proxy() UNLOCK_FUNCTION() {}

        T* operator->()
        {
            return &m_parent.m_obj;
        }

        template <typename Key>
        auto& operator[](Key key)
        {
            return (&m_parent.m_obj)->operator[](key);
        }

        auto begin()
        {
            return (&m_parent.m_obj)->begin();
        }

        auto end()
        {
            return (&m_parent.m_obj)->end();
        }

    private:
        Synced& m_parent;
    };

private:
    mutable Mutex m_mutex;
    T m_obj GUARDED_BY(m_mutex);
};

#define SYNCED_LOCK(synced, name) decltype(synced)::Proxy name{synced}
```

With example usage:

```cpp
Synced<std::vector<int>> synced{1, 2, 3, 4};

{
    SYNCED_LOCK(synced, p);

    p->push_back(5);
}

{
    SYNCED_LOCK(synced, p);

    std::cout << p[0] << std::endl;

    for (auto& i : p) {
        std::cout << i << std::endl;
    }
}
```

We loose the sweet syntax:

```cpp
Synced<std::vector<int>> synced{1, 2, 3, 4};

synced->push_back(5);
```

compared to the other variants, but this would still offer simplifications in higher level code and if you are ok with it, then I am fine.

Reply:
I'm not sure what the differences are between you're quoting above vs what I suggested? They currently look pretty much the same to me? (I was going to check that `begin`/`end`/`op[]` didn't create double locking issues, but you've gotten rid of them already)

I think it'd be worth adding `LIFETIMEBOUND` annotations (from attributes.h) as well. Perhaps you should delete the move constructor for `Synced`?

I agree `synced->push_back(5)` is pretty, but it seems likely to be bug prone to me: consider `it = synced_map->find(key)` -- the lock will be released as soon as the iterator is returned, allowing `it` to be immediately invalidated. So I kind of think we should just count ourselves lucky that we can't make it work and callers have to be explicit about the lock scope...

Reply:
It might be nice to have a `ConstProxy` variant too, so that if you have a `const` reference to an object containing a `Synced<foo>` you can still access have synced access to the object as `const foo`. On the other hand, maybe better to worry about that when an actual use comes up.

Reply:
`Proxy : ::UniqueLock<Mutex>` -- isn't it better to say `private` explicitly? (Guess who just assumed it defaulted to `public` until the compiler told him otherwise...)

Reply:
> I'm not sure what the differences are between you're quoting above vs what I suggested?

Yes, they are pretty much the same. The question was maybe more directed to @ryanofsky - if he is ok with that. I want to avoid pushing another iteration of this PR and it to be shot by somebody shortly after :) If you both agree on the above, then lets do that!

To-polish: LIFETIMEBOUND, delete the Synced move constructor, consider const, explicit inheritance type.

Reply:
If you made `m_obj` and `m_mutex` be protected instead of private, that would allow you to write:

```c++
class MyFoo : public Synced<foo>
{
public:
    int pretty() { LOCK(m_mutex); return m_obj.pretty(); }
};
```

and then do `i = myfoo.pretty();` -- gets you the same end result as the `synced->pretty()` stuff, but only for things you explicitly add a wrapper for, which at least also gives you a chance to make sure it's actually safe to be used in isolation?

Reply:
Ok, I wouldn't add features that are not used anywhere, but `s/private/protected` is not actually more code. Makes sense if somebody wants to inherit it and do more involved things.

To-polish: LIFETIMEBOUND, delete the Synced move constructor, consider const, explicit inheritance type, protected `m_mutex` and `m_obj`.

@ryanofsky, what do you think?

Reply:
> To-polish: LIFETIMEBOUND, delete the Synced move constructor, consider const, explicit inheritance type, protected `m_mutex` and `m_obj`.
> 
> @ryanofsky, what do you think?

All of these tweaks sound good to me.

Generally speaking about the PR as a whole, I'm a skeptical the `Synced` class will be broadly useful, but I think it's a simple enough utility that can be used in a few places without many drawbacks, and that it's worth experimenting with. I was more interested in the class previously when I thought it was a locking pointer enforcing use of an external mutex (that could be used to migrate code away from using `cs_main`), than a locking container that just used an internal mutex. But it seems like a good thing to use in places where it fits.

Reply:
Updated, I only left the `const` stuff away, since it is not needed now and can be added later if needed.

For the usefulness of the `Synced` class, see:
* the second half of https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1169788319
* the second half of https://github.com/bitcoin/bitcoin/pull/26151#pullrequestreview-1123626545
* the commits from this PR that put it into action.

Reply:
Hmm, when used with shared locks, it seems like clang's thread safety analysis doesn't actually prevent you from calling non-const member functions of a `GUARDED_BY` object when only holding the lock in shared mode (it does prevent you from doing `a = b`, just not `a.set(b)`). So if you take a shared lock over a vector, then resize it and invalidate everyone elses iterators, clang thinks it's all fine.

Which I think means that if we wanted to do shared locking more broadly, then this method might be the safest way we have of it: ie have `SYNCED_SHARED(foo,  fooproxy)` take a shared lock and have `*fooproxy` give a `const Foo&` so you can't accidently modify `foo` while you're potentially sharing it with other threads.